vsintegration/src/FSharp.Editor/Diagnostics/UnusedDeclarationsAnalyzer.fs

```diff
@@ -0,0 +1,119 @@
+                match symbolUse.Symbol with
+                | :? FSharpUnionCase when symbolUse.Symbol.IsInternalToProject -> Some symbolUse.Symbol
+                // Determining that a record, DU or module is used anywhere requires
+                // inspecting all their enclosed entities (fields, cases and func / vals)
+                // for usefulness, which is too expensive to do. Hence we never gray them out.
+                | :? FSharpEntity as e when e.IsFSharpRecord || e.IsFSharpUnion || e.IsInterface || e.IsFSharpModule || e.IsClass -> None
+                // FCS returns inconsistent results for override members; we're skipping these symbols.
+                | :? FSharpMemberOrFunctionOrValue as f when f.IsOverrideOrExplicitInterfaceImplementation -> None
+                // Usage of DU case parameters does not give any meaningful feedback; we never gray them out.
+                | :? FSharpParameter -> None
+                | _ when count = 1 && symbolUse.IsFromDefinition && symbolUse.Symbol.IsInternalToProject -> Some symbolUse.Symbol
+                | _ -> None)
+        HashSet(declarations, symbolComparer)
+
+    override __.SupportedDiagnostics = ImmutableArray.Create Descriptor
+
+    override this.AnalyzeSyntaxAsync(document: Document, cancellationToken: CancellationToken) =
+        asyncMaybe {
+            match getProjectInfoManager(document).TryGetOptionsForEditingDocumentOrProject(document) with 
+            | Some options ->
+                let! sourceText = document.GetTextAsync()
+                let checker = getChecker document
+                let! _, _, checkResults = checker.ParseAndCheckDocument(document, options, sourceText = sourceText, allowStaleResults = true)
+                let! symbolUses = checkResults.GetAllUsesOfAllSymbolsInFile() |> liftAsync
+                let declarations = getSingleDeclarations symbolUses
+                if declarations.Count = 0 then return! None
+                else
+                    let! projectCheckResults = checker.ParseAndCheckProject(options) |> liftAsync            
+                    let! allProjectSymbolUses = projectCheckResults.GetAllUsesOfAllSymbols() |> liftAsync
+                    let unusedSymbolUses =
+                        allProjectSymbolUses
+                        |> Array.filter (fun symbolUse -> declarations.Contains symbolUse.Symbol)
+                        |> countSymbolsUses
+                        |> Seq.choose (fun (KeyValue(symbolUse, count)) -> if count < 2 then Some symbolUse else None)
+                    
+                    return 
+                        (unusedSymbolUses
+                         |> Seq.map (fun symbolUse ->
+                             Diagnostic.Create(
+                                 Descriptor,
+                                 CommonRoslynHelpers.RangeToLocation(symbolUse.RangeAlternate, sourceText, document.FilePath)))
+                        ).ToImmutableArray()
```

Comment:
I've seen this is quite a few places - is there nowhere we can have something like:
```f#
module ImmutableArray =
    let make (x : 'a seq) = x.ToImmutableArray ()
```

That way we can do:
```f#
unusedSymbolUses
|> Seq.map (* ... *)
|> ImmutableArray.make
```

Just a thought